### PR TITLE
Fix actor creation hang due to race in SWAP queue

### PR DESCRIPTION
**Pull request by ericl**
## Why are these changes needed?

Currently, test_actor_lifetime_load_balancing will occasionally run much slower (20s+) in raylet mode, and hang forever in direct call mode. This is due to a race condition in the spillback scheduling: if an actor is spilled to another node and back before the creation task is moved out of the swap queue, the creation task gets lost forever.

The fix is simple: don't put it in the swap queue in the first place.

### CODE REVIEW

**Comment by edoakes** [comment]
Deferring to @stephanie-wang on this

**Review by stephanie-wang — APPROVED** [review]

**Comment by stephanie-wang** [comment]
You might want to only do this for direct call tasks, where the workers will be in charge of task resubmissions. If it's done for the raylet codepath too, I think it could end up reopening #4765.

**Comment by ericl** [comment]
Hmm, I would prefer to enable it for both, since it does solve the actor creation hang even in the legacy codepath.

**Comment by ericl** [comment]
If this is a serious concern I can gate it around the actor creation spec is_direct_call flag?

**Comment by stephanie-wang** [comment]
Hmm I guess we'll just try it and see. Hopefully, we can just remove the swap queue fast enough that it won't be an issue.

File: src/ray/core_worker/core_worker.cc [hunk 1]
```diff
@@ -704,9 +704,9 @@ bool CoreWorker::AddActorHandle(std::unique_ptr<ActorHandle> actor_handle) {
           it->second->Reset();
         }
       } else if (actor_data.state() == gcs::ActorTableData::DEAD) {
-        RAY_CHECK_OK(gcs_client_->Actors().AsyncUnsubscribe(actor_id, nullptr));
         // We cannot erase the actor handle here because clients can still
-        // submit tasks to dead actors.
+        // submit tasks to dead actors. This also means we defer unsubscription,
```
Comment (ericl) [inline src/ray/core_worker/core_worker.cc]: I also found this issue debugging a flaky test.

Comment (stephanie-wang) [reply src/ray/core_worker/core_worker.cc]: I'm a bit confused, what is the crash that occurs during the unsubscribe?

**Comment by ericl** [comment]
The crash is when we iterate over actor handles to delete them all in
SetTaskId().

File: src/ray/core_worker/core_worker.cc (hunk 1 above)
Comment (ericl) [reply src/ray/core_worker/core_worker.cc]: The crash is when we iterate over actor handles to delete them all in
SetCurrentTaskId().

**Comment by zhijunfu** [comment]
@ericl It seems like this doesn't solve the issue that #6271 is trying to solve. Please correct me if I misunderstand anything.

Consider the case that a task was forwarded to another raylet, and then was spilled back, and the spillback request arrived earlier than the original forward reply. 

In this case, when the spillback request arrives, since you didn't add it to the swap queue, the spillback task will not be ignored and will get processed correctly, and that will add the task to local queue. But when the original forward reply arrives, it will find the task in the local queue and cancel it. That seems to be the issue this PR doesn't address. In #6271 I used `num_forwards` to check against this.

File: src/ray/core_worker/core_worker.cc (hunk 1 above)
Comment (stephanie-wang) [reply src/ray/core_worker/core_worker.cc]: Sorry I still don't understand completely, why does that crash happen?

Comment (ericl) [reply src/ray/core_worker/core_worker.cc]: You can't AsyncUnsubscribe() twice. It will crash the second time with "subscription not found".

**Comment by ericl** [comment]
> In this case, when the spillback request arrives, since you didn't add it to the swap queue, the spillback task will not be ignored and will get processed correctly, and that will add the task to local queue. But when the original forward reply arrives, it will find the task in the local queue and cancel it. That seems to be the issue this PR doesn't address. In #6271 I used num_forwards to check against this.

How about this: when handling the forward reply, if we find the task is in our local queue again we skip the rest of the handler? Example:
```
  client->ForwardTask(request, [this, on_error, task, task_id, node_id](
                                   Status status, const rpc::ForwardTaskReply &reply) {
      if (local_queues_.HasTask(task)) {
        return;  // It must have been forwarded back to us if it's in the queue again.
      }
```

**Pull request merged**